pysd/py_backend/functions.py

```diff
@@ -1186,6 +1186,93 @@ def random_uniform(m, x, s):
+    It is used in SAMPLE IF TRUE function, to create the proper dimension saved value.
+    
+    Parameters
+    ----------
+    rows: float or xarray.DataArray    
+        Represents the row dimension of the new DataArray
+    cols: xarray.DataArray 
+        Represents the col dimension of the new DataArray
+    initial_value: float or xarray.DataArray 
+        Include the values to initialize the new DataArray
+
+    Returns
+    -------
+    A new DataArray with proper rows and cols coordinates,
+    initialized with initial_value
+    
+    """
+    if(isinstance(initial_value,xr.DataArray)):
+        array = np.array([[initial_value.data[e] for i in range(0, len(rows.values))] for e in range(0,len(cols.values))])
+    elif(isinstance(rows, xr.DataArray)):
+        array = np.array([[initial_value for i in range(0, len(rows.values))] for i in range(0,len(cols.values))])
+    else:
+        array = np.array([initial_value for i in range(0,len(cols.values))])
+        
+    coords = {dim: cols.coords[dim] for dim in cols.dims}
+    dims = cols.dims 
+    if(isinstance(rows, xr.DataArray)):
+        coords.update({dim: rows.coords[dim] for dim in rows.dims})
+        dims += rows.dims
+    return xr.DataArray(data=array, coords=coords, dims=dims)  
+    
+def sample_if_true(time, condition, actual_value, initial_value, var_name):
+    """
+    Implements Vensim's SAMPLE IF TRUE function.
+
+    Parameters
+    ----------
+    condition: bool or xarray.DataArray 
+    actual_value: float or xarray.DataArray 
+        Value to return when condition is true.
+    initial_value: float or xarray.DataArray 
+        Value to return when condition is false.
+    var_name: str
+        Represents the SAMPLE IF TRUE function in the whole model.
+
+    Returns
+    -------
+    float or xarray.DataArray
+        Actual_value when condition is true and saved this value
+        in saved_value dictionary.
+        Returns the last saved value when condicion is false.
+        Saved value is initialized with initial_value in the first step of simulation.
+    """
+    global saved_value
+    t = time()
+
+    if(t==0):
+        if(not(isinstance(condition,xr.DataArray))):
+            saved_value[var_name] = initial_value
+        else:
+            saved_value[var_name] = make_da(actual_value, condition, initial_value)
+
+    if isinstance(condition, xr.DataArray):     
+        if condition.all():
+            for i in range(0,len(saved_value[var_name].values)):
+                saved_value[var_name].values[i]=actual_value
+            return saved_value[var_name]
+        elif not condition.any():
+            return saved_value[var_name]
+
+        for i in range(0, len(condition)):
+            if condition.values[i]:
+                saved_value[var_name][i]=actual_value
+        return xr.where(condition, actual_value, saved_value[var_name])
+
+    if condition:
+        saved_value[var_name] = actual_value
+
+    return saved_value[var_name]
```

Comment:
I suspect that you didn't notice that in the test for the sample_if_true model I indicated that I was working on it.

As this function requires an initial value and information about a saved value, it seems reasonable to implement it as a stateful object. In fact, I have tested an implementation of SampleIfTrue as a Stateful object which requires less code and is more consistent with the PySD structure. 

I will be able to make a PR by next week.

Reply:
Hi @enekomartinmartinez,
I had not seen that you were working on _sample if true_ function until I looked up if there was any test for this function. By then, I had had a good part of implementation done yet. 

Maybe, they could add my function to the repo, because this implementation for _sample if true_ works, and then, when you make your pull request, change and improve the _sample if true_ function.

In addition, I take the opportunity to ask you about what are you working on? Or what can I work on to make progress?

Reply:
Hi @marrobl,
Sorry for taking so long to answer.
The problem of changing the sample_if_true function by a stateful object SampleIfTrue is that we will make a non-backward compatible version. We need also to make the initialization of the function solid, in the case of using a Stateful object this is simple as we have the initialize method. Note that your function may not be well initialized if initial_time is not 0.

You can incorporate the SampleIfTrue inherited from Satetful class, use a similar builder like the one used for Integ or Delay as it makes it possible to work with subscripted variables. The initialization of the object could be similar to the one done with Integ.  You can create a call method that evaluates the condition with the current if_then_else function and updates the state if needed. As was done with the Initial class, it may be necessary to add a ddt equal to 0 and pass when update is call.

Reply:
Hi @enekomartinmartinez, 
Okey, I'm on it. I'm going to try to develop _sample if true_ function the way you have told me. Hope to have it finished soon so I can include in this pull request

Reply:
Hi @marrobl, the SampleIfTrue class looks very nice :)
Yes, _if_then_else_ takes the _val_if_true_ and _val_if_false_ as functions just to use lazy evaluation as VENSIM does. Have you tried instead using a lambda function in _self.state_ this will solve the problem without needing to create a new function:
```
self.state = if_then_else(self.condition(), self.actual_value, lambda: self.state)
```
if you are able to check that I think we can already merge this pull request.

Reply:
You're right @enekomartinmartinez! Thanks, I did not realize how to 'cast' a simple value to a function.